Improve NVCC & Kokkos support, test on CI

nilsvu opened the pull request:
## Proposed changes

A few things that allowed me to compile with Kokkos.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Support Kokkos nvcc_wrapper

nilsvu committed:
Fix disabling warnings

nilsvu committed:
Enable constexpr in CUDA

nilsvu committed:
Fix a warning in elliptic code

Template parameter unused

nilsvu committed:
Clean up Overloader.hpp a bit

nilsvu committed:
Support nvcc in Test_TMPLDocumentation.cpp

Relax some tests so they pass for both
std::true_type and tmpl::true_type
because Brigand has different code paths.

nilsvu committed:
Test compiling with nvcc on CI

nilsvu committed:
Move CUDA warnings config to EnableWarnings.cmake

This enables the config when compiling with nvcc, even
if Kokkos is not enabled.

nilsvu committed:
Fix an issue with nvcc and Blaze 3.8.2

nilsvu committed:
Fix boost variadic macros with nvcc

nilsvu committed:
Disable two tests for nvcc

nvcc can't handle TaggedVariant at the moment
and runs out of memory for Test_TimeDerivative.cpp

nilsvu committed:
Suppress more CUDA warnings

nilsvu committed:
Fix a warning in pypp

nilsvu committed:
Fix a missing namespace

nilsvu committed:
Remove `source_location` from a test

nvcc can't handle it, and it seems helpful but not essential for the test.
See upstream issue: https://forums.developer.nvidia.com/t/internal-error-assertion-failed-compare-expressions/303471

nilsvu committed:
Fix an elliptic utility to work with nvcc

nilsvu committed:
Get nvcc to compile elliptic executables

nilsvu committed:
Move NewtonianEuler VolumeTermsInstantiation

Uses excessive memory when compiler with nvcc. Moving
to the executable and compiling only the needed dimension.
Also reduce EvolveNewtonianEuler3D input file priority so it
isn't compiled for nvcc on CI.

nilsvu committed:
Fix a warning in binary domain creator

nilsvu committed:
Init and finalize kokkos in executables

nilsdeppe commented on cmake/SetupKokkos.cmake:
```diff
@@ -43,24 +43,26 @@ if(SPECTRE_KOKKOS)
       AND CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-    set_property(TARGET Kokkos::kokkos
+    # Not using the 'Kokkos::kokkos' target because that's an alias to 'kokkos',
+    # and we can't set properties on alias targets
```
Do you know if this is kokkos version dependent? Because this had worked for me previously. We might need to pin to kokkos versions like we do for blaze anyway. I'm not asking you to change anything, just curious if you know. "No" is totally fine :)

nilsvu replied:
Looks like the `Kokkos::kokkos` target can either be an alias or not, depending on how it was installed (pointing to source tree vs installed on system), so I think we have to support both. I'm thinking, can we just remove these two `set_property` calls? The first I already moved to `EnableWarnings.cmake` anyway, and the second (the one with `-Wno-unknown-cuda-version`) we could also either move to `EnableWarnings.cmake` or delete altogether. The mapping of clang to cuda versions is not so straightforward anyway, so we might as well emit the warning if the versions are marked incompatible. What do you think?

nilsdeppe commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -575,41 +575,45 @@ assert_same<tmpl::reverse_range<size_t, 7, 7>, tmpl::list<>>();
 // [tmpl::all:inhomogeneous]
-assert_same<tmpl::all<List1<std::true_type, tmpl::true_type>, tmpl::_1>,
-            tmpl::false_type>();
+#ifndef __CUDACC__
```
Do you know why some of these fail with nvcc? It looks like a lot of them work from the any section. I'm worried we use these features and nvcc fails.

Would it make sense to move the ifndef outside the doxygen snippet statements?

nilsvu replied:
Yes I know what's happening here. Brigand has a special code branch for nvcc (here: https://github.com/sxs-collaboration/spectre/blob/develop/external/brigand/include/brigand/brigand.hpp#L951). Unfortunately it behaves slightly different to the standard code branch, in that it returns `std::true_type` where the standard branch returns `tmpl::true_type`. Therefore I put a warning in the docs saying the behavior may depend on the compiler. As long as we always use `::value` this should not matter, and I checked and found that we never use `tmpl::true_type` or `tmpl::false_type` in the code, so we should be fine.

nilsvu replied:
Note that I also tried to remove the special code branch in Brigand for nvcc and it compiled fine, so maybe this was only necessary in Brigand for an old version of nvcc. But it was easier to modify our tests since we never rely on `std::true_type` vs `tmpl::true_type` in our code anyway, than to change upstream Brigand or patch it.

nilsvu replied:
I kept the ifndef inside the snippet to emphasize in the docs that it can be compiler-dependent

nilsdeppe commented on cmake/SetupKokkos.cmake:
```diff
@@ -62,4 +62,18 @@ if(SPECTRE_KOKKOS)
       )
```
Remove "fixup disable another test" from commit message?

nilsdeppe commented on tests/Unit/DataStructures/CMakeLists.txt:
```diff
@@ -47,6 +46,11 @@ set(LIBRARY_SOURCES
   Test_VectorImplTestHelper.cpp
   )
 
+# The following test is disabled for nvcc because nvcc can't handle it.
```
maybe specify what fails? E.g. it just runs out of RAM?

nilsdeppe commented on tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/CMakeLists.txt:
```diff
@@ -30,6 +29,11 @@ set(LIBRARY_SOURCES
   Test_TimeDerivativeTerms.cpp
   )
 
+# The following test is disabled for nvcc because nvcc can't handle it.
```
maybe specify what fails? E.g. it just runs out of RAM?

nilsdeppe reviewed (changes requested):
LGTM, just a few questions!

nilsvu committed:
Remove a kokkos config specific to clang&cuda versions

The mapping between compatible clang and cuda versions
is too complicated to reproduce in our cmake config, so
just emit the warning if they are deemed incompatible.

nilsdeppe reviewed (dismissed):
LGTM! Please squash. Thanks for doing this!!
